File: 1424_diagonal_traverse_ii/Solution.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& nums) {
        int maxLength = 0;
        std::map<int, std::vector<int>> result;
        
        for(int y = nums.size() - 1; y >= 0; y--){
            for(int x = 0; x < nums[y].size(); x++){
                if(result.count(y+x) == 0){
                   result.insert({y+x, std::vector<int>()}); 
                }
                result.at(y + x).push_back(nums.at(y).at(x));
                if(y+x > maxLength){
                    maxLength = y+x;
                }
            }
        }
        
        std::vector<int> ans;
        
        for(int i = 0; i <= maxLength; i++){
            for(int j = 0; j < result.at(i).size(); j++){
                ans.push_back(result.at(i)[j]);
            }
        }
        
        return ans;
    }
};

File: 1424_diagonal_traverse_ii/Solution.java
class Solution {
    public int[] findDiagonalOrder(List<List<Integer>> nums) {
        int returnLength = 0;
        int maxLength = 0;
        Map<Integer, List<Integer>> result = new HashMap<>();
        
        for (int y = nums.size() - 1; y >= 0; y--) {
            for (int x = 0; x < nums.get(y).size(); x++) {
                result.putIfAbsent(y + x, new ArrayList<>());
                result.get(y + x).add(nums.get(y).get(x));
                maxLength = Math.max(maxLength, y + x);
                returnLength++;
            }
        }
        
        int[] ans = new int[returnLength];
        int i = 0;
        
        for (int key = 0; key <= maxLength; ++key) {
            List<Integer> value = result.get(key);
            if (value == null) continue;
            for (int v : value) ans[i++] = v;
        }
        return ans;
    }
}
